### PR TITLE
[MINOR][SQL][DOC] Fix some format issues in Dataset API Doc

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

fix listing up format issues in Dataset API Doc (scala & java)

### Why are the changes needed?

improve doc

### Does this PR introduce any user-facing change?

yes, API doc changing 

### How was this patch tested?

no

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc: @maropu @HyukjinKwon @cloud-fan @dongjoon-hyun, thanks in advance for reviewing this.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -1868,9 +1868,9 @@ class Dataset[T] private[sql](
  /**
   * Define (named) metrics to observe on the Dataset. This method returns an 'observed' Dataset
   * that returns the same result as the input, with the following guarantees:
-  * - It will compute the defined aggregates (metrics) on all the data that is flowing through the
+  *  - It will compute the defined aggregates (metrics) on all the data that is flowing through the
   *   Dataset at that point.
-  * - It will report the value of the defined aggregate columns as soon as we reach a completion
+  *  - It will report the value of the defined aggregate columns as soon as we reach a completion
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: To make completely fix it, you gotta use HTML tag to render properly in both javadoc and scaladoc. See https://github.com/apache/spark/blob/9dedfeffaa27a00f719b5a50ce1692f6a0ad665d/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L527-L536 as an example.

You can simply revert this change too here.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -2498,6 +2498,13 @@ class Dataset[T] private[sql](
    *
    * Use [[summary]] for expanded statistics and control over which statistics to compute.
    *
+   * NOTES:
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: You can use `@note`. It works fine in both javadoc and scaladoc.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 3]
```diff
@@ -2498,6 +2498,13 @@ class Dataset[T] private[sql](
...
+   *  - String columns are countable, so count is accurate
+   *  - String columns are comparable, so min/max are accurate
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: I don't understand what both mean. Strings are already countable and comparable.

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: yes, maybe I just over explained

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 4]
```diff
@@ -2498,6 +2498,13 @@ class Dataset[T] private[sql](
...
+   *  - String columns will be implicitly cast to double when perform arithmetic operations, so the
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: `when perform arithmetic operations` -> `when performing arithmetic operations`

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Why do we talk about arithmetic operations here?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 5]
```diff
@@ -2512,12 +2519,12 @@ class Dataset[T] private[sql](
...
-   * - stddev
-   * - min
-   * - max
-   * - arbitrary approximate percentiles specified as a percentage (eg, 75%)
+   *  - count
+   *  - mean
+   *  - stddev
+   *  - min
+   *  - max
+   *  - arbitrary approximate percentiles specified as a percentage (eg, 75%)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: ditto about listing up. This listing up doesn't work already:
Screen Shot 2019-12-17 at 8 47 13 PM

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: image
one more space is enough for listing up

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Yeah, but it doesn't work in javadoc. I meant this is right fix for scaladoc but not for javadoc.

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: oh I missed java doc

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 6]
```diff
@@ -2498,6 +2503,10 @@ class Dataset[T] private[sql](
    *
    * Use [[summary]] for expanded statistics and control over which statistics to compute.
    *
+   * @note String columns will be implicitly cast to double when performing operations like mean
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: I think this documentation should rather go to `functions.scala` or the SQL documentation under `ExpressionDescription`. This seems like not only specific to this API alone. You can remove this here and do it separately in a separate PR.

**Review by HyukjinKwon — APPROVED** [review]
Seems fine except https://github.com/apache/spark/pull/26922/files#r358762281

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 6 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: This method accept numeric and string, I guess this note help us understand the string stat output. For sepecific functions, the implicit casting happens not only for string only.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: But this applies to all other stddev and mean functions too:

```scala
scala> sql("select mean('1.0')").show()
+------------------------+
|avg(CAST(1.0 AS DOUBLE))|
+------------------------+
|                     1.0|
+------------------------+
```

I would locate the note in a better place to deduplicate the documentation.

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: all right

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 7]
```diff
@@ -450,12 +450,14 @@ class Dataset[T] private[sql](
...
    * Returns a new Dataset where each record has been mapped on to the specified type. The
    * method used to map columns depend on the type of `U`:
-   *  - When `U` is a class, fields for the class will be mapped to columns of the same name
-   *    (case sensitivity is determined by `spark.sql.caseSensitive`).
-   *  - When `U` is a tuple, the columns will be mapped by ordinal (i.e. the first column will
-   *    be assigned to `_1`).
-   *  - When `U` is a primitive type (i.e. String, Int, etc), then the first column of the
-   *    `DataFrame` will be used.
+   * <ul>
+   *   <li>When `U` is a class, fields for the class will be mapped to columns of the same name
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Nit: close `<li>` tag

Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: BTW... doesn't scaladoc interpret markdown tags? Is it just that these didn't have a blank line above or something? I don't know it well.

**Review by srowen — COMMENTED** [review]
This is fine; ar

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 7 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: nice catch!thanks! the scaladoc supports wiki-style syntax and HTML, not markdown.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 8]
```diff
@@ -2511,13 +2516,14 @@ class Dataset[T] private[sql](
...
-   * - min
-   * - max
-   * - arbitrary approximate percentiles specified as a percentage (eg, 75%)
+   * <ul>
+   *   <li>count</li>
+   *   <li>mean</li>
+   *   <li>stddev</li>
+   *   <li>min</li>
+   *   <li>max</li>
+   *   <li>arbitrary approximate percentiles specified as a percentage (eg, 75%)</li>
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: nit `eg,` => `e.g.,`

**Review by maropu — APPROVED** [review]
Looks nice, thanks @yaooqinn

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 7 above)
Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Oh OK. Maybe I'm crazy but I thought `- foo` style lists work, but remember that sometimes they don't unless there's a blank line above it. I'm just wondering if that's the issue here. Or if it's not -- there are many more instances like this in the code, no? let's fix more of them if it's necessary

**Comment by HyukjinKwon** [comment]
Merged to master and branch-2.4.

**Comment by yaooqinn** [comment]
Thanks for merging and reviewing. close via d38f8167483d4d79e8360f24a8c0bffd51460659
